Cache Canary camera image

0bmay opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add cached image for camera so a new image is not retrieved upon every refresh

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: [PR 73922](https://github.com/home-assistant/core/pull/73922)
- Link to documentation pull request: 

with py-canary 0.5.3 the updated image will refresh every time an image is requested flooding the Canary servers to get a new image every 10 or so seconds.  This will grab an image and hold it until the next reload (next pull request will use canary entries api to update the image cache when a new entry is updated)

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

0bmay committed:
Cache camera image

Cache camera image so a new image isn't generated each call.
Adds debug logging

frenck commented:
Can you rebase this PR now #73922 has been merged?

0bmay commented:
Will attempt.. (I usually mess this part up)

0bmay commented:
appears to have worked.. relatively new to git rebase..

0bmay commented:
local tests work fine for this branch after rebase.. 🤷

frenck commented:
Yes, it seems we are currently experiencing issues with our CI. At this point, we suspect it to be an issue at GitHub. Will get back on that.

frenck commented:
Your rebase timing is amazing :D Just merge the actual fix :)

0bmay commented:
I found it not too long ago and was following along... :)

emontnemery commented on homeassistant/components/canary/camera.py:
```diff
@@ -125,17 +128,36 @@ async def async_camera_image(
+            _LOGGER.debug("Grabbing a live view image from %s", self.name)
+            await self.hass.async_add_executor_job(self.renew_live_stream_session)
+
+            if self._live_stream_session is None:
```
```suggestion
            if (live_stream_session := self._live_stream_session) is None:
```

emontnemery commented on homeassistant/components/canary/camera.py:
```diff
@@ -125,17 +128,36 @@ async def async_camera_image(
+            if self._live_stream_session is None:
+                return None
+
+            live_stream_url = self._live_stream_session.live_stream_url
+            if not live_stream_url:
```
```suggestion
            if not (live_stream_url := live_stream_session.live_stream_url):
```

emontnemery commented on homeassistant/components/canary/camera.py:
```diff
@@ -125,17 +128,36 @@ async def async_camera_image(
+            if not live_stream_url:
+                return None
+
+            if self._image is None and live_stream_url:
```
This check must be useless?
```suggestion
```

allenporter replied:
I think you should accept this suggestion and add it back in the future PR.

emontnemery commented on homeassistant/components/canary/camera.py:
```diff
@@ -125,17 +128,36 @@ async def async_camera_image(
+                )
+                _LOGGER.debug("Stopped live session from %s", self.name)
+
+        return self._image
```
Shouldn't `self._image` be invalidated after some time?

0bmay replied:
Good eye.. yes, the image will be invalidated/updated in the next PR that comes after this one [PR 73925](https://github.com/home-assistant/core/pull/73925).  This PR is 2/5 that were split out from 1 larger PR at @frenck's request.

allenporter replied:
I had a similar thought as @emontnemery  -- Definitely makes sense to split into smaller PRs! but it still seems like there should be some kind of invalidation consideration in the same PR where the cache is added e.g. checking for the URL to change?  or invalidating after some expiration time.

I had a look at the other PRs and it's a little unclear about the order they should be reviewed given they have overlapping/duplicate changes.

0bmay replied:
Thank you @allenporter, as I will have to go the multiple PR route again when adding in 2FA for canary after I finish testing that code. I don't want to create another mess at that time.  I will work on adding an expiration to the cache.

0bmay replied:
marked the other pull requests as draft requests.. will enable when needed.

0bmay committed:
Apply suggestions from code review

code compression with walrus operator

Co-authored-by: Erik Montnemery <erik@montnemery.com>

0bmay committed:
fix after walrus operator suggested tweak

fully use the live_stream_session variable in async_camera_image

0bmay committed:
Invalidate cached image after 15 minutes

requested code change; invalidate cached image

allenporter reviewed (changes requested):
One comment to +1 emontnemery's comment, but otherwise looks good to me.

0bmay committed:
Removed unnecessary if statement

based on code review

allenporter commented on homeassistant/components/canary/camera.py:
```diff
@@ -105,6 +108,8 @@ def __init__(
             name=device.name,
         )
+        self._image: bytes | None = None
+        self._expires_at = dt_util.utcnow() - FORCE_CAMERA_REFRESH_INTERVAL
```
Is this 15 minute ago?

allenporter replied:
I see, it's arbitrary since there is no image set.   The math here i think is not needed. Maybe just use `datetime.min` so it might be more clear its a placeholder small value?

0bmay replied:
set to `dt_util.utcnow()` no math needed and works the same as before.

allenporter commented on homeassistant/components/canary/camera.py:
```diff
@@ -125,17 +130,40 @@ async def async_camera_image(
+        if self._expires_at <= utcnow:
+            _LOGGER.debug("Expiring cached image from %s", self.name)
+            self._image = None
+            self._expires_at = FORCE_CAMERA_REFRESH_INTERVAL + utcnow
```
I realize its somewhat arbitrary when  you set the expiration time, but logically to me it makes the most sense to set `_expires_at` in the same block you set `_image`.

I would suggest that it might be easiest to just inline the `expires_at` check  on the same line as the `if self._image is None:` check then also update the expiration time after grabbing the image.  In general, I do like separation of responsibilities, but i think here its easier to reason about all in one function.

0bmay replied:
done; now has the added benefit of not clearing the image prior to getting a new one. This way if there is an issue getting the new image it will not display a broken image.

0bmay committed:
Image capture flow updates

now sets the image expiration upon getting an updated image
updates the cache image only when a new image is captured

allenporter commented on homeassistant/components/canary/camera.py:
```diff
@@ -125,17 +133,33 @@ async def async_camera_image(
-            height=height,
-        )
+        utcnow = dt_util.utcnow()
+        if self._expires_at <= utcnow:
```
Do you want to retry every time if the image fails to fetch? e.g. 

```suggestion
        if not self._image or self._expires_at <= utcnow:
```

0bmay replied:
the only time `self._image` is None is right after `init()` and before a successful image retrieval. `self._expires_at` is only updated after a successful image retrieval, so if there is an error, the next time an image is requested, `self._expires_at` would be less than 'now' and an image retrieval attempt would be made again.

With that said, there's no harm in adding it, but with other requests to simplify if statements, I feel this would be redundant.

0bmay replied:
Also, before I forget, thank you for your feedback. 😀

0bmay commented:
@emontnemery can you please review so this fix can merge in please and thank you.

emontnemery reviewed (approved):
Thanks, @0bmay 👍
